### ISSUE TITLE
Multithreading

### ISSUE THREAD

**Issue by ethicalhack3r** [postmodern/spidr#26]
Hi there,

I was wondering if it would be possible to multithread the spidr gem? I don't know much about multithreading in ruby, but I believe only Ruby 1.9.x is able to do so?

I had a look through the source but couldn't find where the spidr gem makes its http requests.

Maybe something like Typhoeus can be used?! (http://rubygems.org/gems/typhoeus)

Thanks,
Ryan

**Comment by dadamschi** [comment]
What happened with this request?

**Comment by postmodern** [comment]
I don't have the time currently to work on such a large feature.

**Comment by ZeroChaos-** [comment]
been a year, any chance you have time to work on such a feature now? :-)

**Comment by fuzzygroup** [comment]
I've written more than a crawler or N in my career and if you didn't make it multi-threaded from the start, it is damn hard to do so in retrospect.  Now, that said, I think the overall goal here is throughput rather than threads.  If the discovered urls can be surfaced to an overall queue (Redis or SQS) then that would change the equation because rather than threads you simply run more instances (or containers) of Spidr and let the queue handle distribution of work across N copies.

Thoughts?

**Comment by postmodern** [comment]
A distributed Spidr is a little out of scope, or at least further down the road.

Multi-threading here is mainly to address blocking I/O when waiting on responses to come back from the HTTP Sessions. Luckily, [net-http-persistent](https://github.com/drbrain/net-http-persistent#readme) is already thread aware. We'd just need to replace the spidering loop with a producer/consumer thread pool. Each thread would have it's own session cache via net-http-persistent, would dequeue URLs, and enqueue the responses/Pages. All additional logic with headers and parsing HTML would still be done in the main thread, to avoid additional Mutex complexity. There's probably other hidden work and locking issues hidden in the details.

**Comment by vwochnik** [comment]
+1, a producer/consumer for the requests would be awesome! I really like the interface of your library by the way.

**Comment by dadamschi** [comment]
I don't understand "a producer/consumer for the requests"...


==================
David Adams
dadams.chi@gmail.com

**Comment by vwochnik** [comment]
I mean a producer/consumer pattern where one a thread pool of worker threads that do the requesting are connected to the main thread with queues, like a manufacturing band.

The main thread puts all requests that it wants to have resolved in a queue and any worker thread can pick the task from the queue, do the request, and put it inside the finished responses queue which is being read by the main thread. In this way, the main thread does not do any requesting, i.e. blocking activity, itself which will lead to a speedup.
